// @prequest/helper/dist/helper.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var qs = require('qs');
var x = require('@xdoer/x');

class PQError extends Error {
  constructor(opt) {
    super();
    this.opt = opt;
    const {
      code,
      config
    } = this.opt || {};
    this.code = code;
    this.config = config;
  }

}

(function (ErrorCode) {
  ErrorCode["timeout"] = "timeout";
  ErrorCode["abort"] = "abort";
  ErrorCode["common"] = "common";
})(exports.ErrorCode || (exports.ErrorCode = {}));

function createError(code, message, opt) {
  const error = new Error(message);
  return enhanceError(error, code, opt);
}
function enhanceError(error, code, opt) {
  error.code = code;
  error.config = opt;
  return error;
}

const baseOption = {
  path: '/',
  method: 'GET'
};
function createRequestUrl(req) {
  const {
    baseURL,
    path,
    params
  } = req;
  let url = req.url || '';

  if (x.isAbsoluteURL(path)) {
    url += path;
  } else {
    if (baseURL) url += baseURL;
    if (path) url += path;
  }

  if (params && !x.isEmpty(params)) url += `?${qs.stringify(params)}`;
  return url;
}
function requestId(options) {
  return createRequestUrl({ ...options,
    params: {}
  });
}

exports.PQError = PQError;
exports.baseOption = baseOption;
exports.createError = createError;
exports.createRequestUrl = createRequestUrl;
exports.enhanceError = enhanceError;
exports.requestId = requestId;
