// @prequest/miniprogram-addon/dist/miniprogram-addon.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@prequest/core');
var helper = require('@prequest/helper');

function adapter(request) {
  return opt => {
    return new Promise((resolve, reject) => {
      const finalOption = opt || {};
      const url = helper.createRequestUrl(finalOption);
      const {
        getNativeRequestInstance,
        cancelToken,
        ...rest
      } = finalOption;
      const instance = request({ ...rest,
        url,
        success: resolve,
        fail: reject
      });

      if (cancelToken) {
        cancelToken.promise.then(cancel => {
          if (!instance) return; // 如果支持取消方法

          if (instance.abort) return instance.abort(); // 如果不支持，则直接抛出错误

          reject(helper.createError(helper.ErrorCode.abort, cancel, opt));
        });
      }

      getNativeRequestInstance === null || getNativeRequestInstance === void 0 ? void 0 : getNativeRequestInstance(instance);
    });
  };
}

function createUpload(request, options) {
  return core.PreQuest.create(adapter(request), options);
}
function createDownload(request, options) {
  return core.PreQuest.create(adapter(request), options);
}

Object.defineProperty(exports, 'PreQuest', {
  enumerable: true,
  get: function () {
    return core.PreQuest;
  }
});
exports.adapter = adapter;
exports.createDownload = createDownload;
exports.createUpload = createUpload;
