// @dcloudio/uni-mp-vite/dist/plugins/runtimeHooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uniRuntimeHooksPlugin = void 0;
const uni_shared_1 = require("@dcloudio/uni-shared");
const uni_cli_shared_1 = require("@dcloudio/uni-cli-shared");
function uniRuntimeHooksPlugin() {
    return {
        name: 'uni:mp-runtime-hooks',
        enforce: 'post',
        async transform(source, id) {
            if (!(0, uni_cli_shared_1.isUniPageSfcFile)(id)) {
                return null;
            }
            if (!source.includes('_sfc_main')) {
                return null;
            }
            const matches = source.match(new RegExp(`(${Object.keys(uni_shared_1.MINI_PROGRAM_PAGE_RUNTIME_HOOKS).join('|')})`, 'g'));
            if (!matches) {
                return null;
            }
            if (matches.includes('onShareTimeline')) {
                matches.push('onShareAppMessage');
            }
            const hooks = new Set(matches);
            let flag = 0;
            for (const hook of hooks) {
                flag |= uni_shared_1.MINI_PROGRAM_PAGE_RUNTIME_HOOKS[hook];
            }
            return {
                code: source + `;_sfc_main.__runtimeHooks = ${flag};`,
                map: { mappings: '' },
            };
        },
    };
}
exports.uniRuntimeHooksPlugin = uniRuntimeHooksPlugin;
